Fix import leaks

rokm opened the pull request:
Fix import leaks that happen when`PyInstaller.utils.hooks.get_module_file_attribute` or `PyInstaller.utils.hooks.is_package` is called with a sub-package or a sub-module name, which causes the `pkgutil.find_loader` to import the parent package.

In addition, several other hookutils functions have been ported to the new `PyInstaller.isolated` framework. The most notable is `get_module_attribute` which now returns actual attribute values (in their original type) rather than their string representations.

As we now manage to completely avoid importing collected packages in the main process, the binary dependencies search need to be performed in an isolated subprocess that first imports all collected packages (so that their imports properly set up search paths). Instead of importing just packages, we should perhaps be importing all collected modules in case some of those are stand-alone and contain search-path initialization themselves.

rokm committed:
hookutils: get_module_file_attribute: fix import leaks

When pkgutil.find_loader() in get_module_file_attribute() is called
with a sub-package or a sub-module name, it leaks an import of the
corresponding parent package(s).

Therefore, if the given module/package name contains a dot, we always
need to use subprocess isolation. Instead fo exec_statement, we now
use the new PyInstaller.isolated framework.

Inside the subprocess-isolated check, we need to first perform the
pkgutil-based check, as it manages to return file paths for packages
and modules that are otherwise un-importable (e.g., a C-extension
module with missing dependencies, such as PyQt5.QtMultimedia with
missing pulseaudio libs on linux).

We also import pkgutil locally, so we can eventually get rid of
module-level import, which will help us ensure that all uses of
pkgutil are properly isolated.

rokm committed:
hookutils: qt: port the QLibraryInfo retrieval to PyInstaller.isolated

rokm committed:
building: isolate the search for dynamic libraries

Isolate the search for dynamic libraries in a subprocess, and have
it import all collected packages in order to properly set up
library search paths. This functionality is implemented as a
module-level helper function that is executed via the new isolated
framework.

We need a stand-alone helper function because one of the tests needs
to monkey-patch it in order to inject its (subproess-)local override
of the getImports() function.

rokm committed:
tests: fix test_hookutils.data_lists fixture

Contrary to the comments, prepending sys.path with the path to
the test package is not useful for subprocesses. Instead, we
need to modify PyInstaller.config.CONF['pathex']; same as already
done for the mod_list fixture.

rokm committed:
ci: install libpulse0 and libpulse-mainloop-glib0 on ubuntu runners

These are required by QtMultimedia, and their absence is causing
import error when trying to import PyQt5.QtMultimedia.

This in turn causes test_PyQt5_Qt to fail when trying to run a hook
for PyQt5.QtMultimedia, because add_qt_dependencies() calls
get_module_file_attribute(), and we have modified the behavior to
immediately use subprocess-isolated import attempt. This behaves
differently from pkgutil.get_loader().get_filename() which
apparently returned file path to package even if the latter cannot
be imported.

So while ultimately the subprocess-isolated codepath of
get_module_file_attribute() should be fixed to restore old behavior,
we should probably ensure that our CI does satisfy the dependencies
of the Qt-based packages it uses...

rokm committed:
tests: fix test_regression::test_issue_5131

This test monkey-patches bindepend.getImports() function to ignore
errors on invalid binary file that the test uses. However, as we
now process binaries in an isolated subprocess, this change is not
progpagated there. So we need to provide getImports() override both
for the test's main process, as well as override the implementation
of build_main.find_binary_dependencies() to inject getImports()
override in the isolated subprocess.

rokm committed:
hookutils: can_import_module: port to PyInstaller.isolated

rokm committed:
hookutils: get_module_attribute: port to PyInstaller.isolated

rokm committed:
hookutils: is_package: prevent import leaks

The calls to this function leak imports when called with sub-module
or sub-package name (because parent is imported by pkgutil). Isolate
the check accordingly.

rokm committed:
hookutils: remove module-level import of pkgutil

All pkgutil imports are now localized, and all uses of pkgutil
functions are now properly isolated where there is a risk of
import leak.

rokm committed:
hooks: django: remove eval() around retrieved module attribute

The utils.hooks.get_module_attribute() function now returns module
attributes in their native types instead of string representation.

rokm committed:
tests: qt: remove Qt pre-import hack

With import leaks fixed and corresponding codepaths isolated in
subprocesses, we should now be able to run the tests without
ensuring that all Qt-based packages are imported before the
tests start.

rokm committed:
tests: qt: remove PATH manipulation work-around on Windows

With import leaks fixed, there should now be no need for cleaning
up PATH on Windows to ensure that only one Qt-based package is in
the path.

rokm committed:
hookutils: qt: fix DLL search path for PySide2 and PySide6

The DLL search path used to resolve import names to fullpaths on
Windows differs between PyQt5/6 and PySide2/6. For the former,
the existing BinariesPath is correct, but for the latter, we need
to use PrefixPath.

Alternatively, we could do away with giving extra search path to
bindepend.getfullnameof(), and instead perform the analysis of
Qt modules' DLL dependencies in an isolated subprocess and with
the main Qt package imported.

But at least for PyPI wheels, giving the search path seems to be
sufficient...

bwoodsend commented:
I notice that
\> `PyInstaller/hooks/hook-pandas.io.formats.style.py` calls
\> `collect_data_files('pandas.io.formats')` which calls
\> `get_package_paths("pandas.io.formats")` which calls
\> `get_module_file_attribute("pandas.io.formats")` which will import `pandas.io` in the master process only for monkeypatch to remove it again. So this PR should finally put that pandas CI failure to rest.

rokm commented:
I've cleaned up the commit history a bit to squash multiple iterations/attempts at fixing a specific part of the code into a single commit, and rebased the branch on top of `develop`.

bwoodsend commented on PyInstaller/building/build_main.py:
```diff
@@ -119,6 +119,39 @@ def discover_hook_directories():
+    Find dynamic dependencies (linked shared libraries) for the provided list of binaries.
+
+    Before scanning the binaries, the function imports the packages from provided list of packages to import, to ensure
+    that library search paths are properly set up (i.e., if a package sets up search paths when imported9. Therefore,
```
I'm guessing you meant:
```suggestion
    that library search paths are properly set up (i.e., if a package sets up search paths when imported). Therefore,
```

rokm replied:
Yeah. Fixed.

bwoodsend commented on PyInstaller/building/build_main.py:
```diff
@@ -482,8 +515,19 @@ def assemble(self):
+        #
+        # Up until this point, we did very best not to import the packages into the main process. However, a package
+        # may set up additional library search paths during its import (e.g., by modifying PATH or calling the
+        # add_dll_directory() function on Windows, or modifying LD_LIBRARY_PATH on Linux). In order to reliably
+        # discover dynamic libraries, we therefore require an environment with all packages imported. We achieve that
+        # by gathering list of all collected packages, and spawn an isolated process, in which we first import all
+        # the packages from the list, and then peform search for dynamic libraries.
```
Does this not put us back into the danger zone that is multiple Qt variants being loaded in the same process leading to the wrong DLLs being collected?

rokm replied:
It does. But if we collect multiple variants of Qt-based packages for some reason (e.g., `Qt5Agg` `matplotlib` backend in an environment where we have both PyQt5 and PySide2 installed), we are risking undefined behavior in other places as well: 
* only one set of shared libs will end up collected in `sys._MEIPASS`
* multiple rthooks are executed at run-time, so QT_PLUGIN_PATH and QML2_IMPORT_PATH are set to the value of  the last-executed qthook

I suppose the practical issues stemming from this are now somewhat mitigated by the fact that both packages are using Qt 5.15.2, so those shared libraries and plugins are (at least in theory) interchangeable.

rokm replied:
So I think this problem existed before this PR (because hooks ended up importing Qt packages during modulegraph analysis) and continues to exist in it (although admittedly, it might manifest itself differently, as the order of imports is very likely to change).

One way to solve it could be implementing some sort of mutual-exclusion mechanism for groups of packages: 
in a post-processing step (after modulegraph analysis but before binary dependency analysis), we'd go over collected packages, and if a higher-ranking package from a group is to be collected, it automatically suppresses other packages from that group.

bwoodsend commented on tests/unit/test_hookutils.py:
```diff
@@ -329,7 +329,9 @@ def _sort(sequence):
+    # Add path with 'hookutils_files' module to ``sys.path`` (so analysis in the main process can find it),
+    # and to ``pathex`` (so subprocess-isolated code can find it).
+    monkeypatch.setattr('PyInstaller.config.CONF', {'pathex': [TEST_MOD_PATH]})
     monkeypatch.syspath_prepend(TEST_MOD_PATH)
```
Doesn't that mean that this `syspath_prepend()` line can be removed then?

bwoodsend replied:
Or is `sys.path` needed for non subprocesses?

rokm replied:
I left it in just in case it might be required by main process, but now I'm not sure it is actually required anywhere in the test. The main idea was to have `data_lists` fixture do the same kind of environment modifications as the `mod_list`  from the same file.

bwoodsend commented:
[That CI failure](https://github.com/pyinstaller/pyinstaller/pull/6171/checks?check_run_id=3506303684) is new which is odd because the pytz hook and test look about as vanilla as they get..

rokm commented:
> [That CI failure](https://github.com/pyinstaller/pyinstaller/pull/6171/checks?check_run_id=3506303684) is new which is odd because the pytz hook and test look about as vanilla as they get..

Yeah, the log suggests that the crash was caused by a timeout in the `matplotlib` hook (isolated retrieval of the data directory), which is really odd, as I don't see why that hook would be run for the pytz test...

bwoodsend commented:
[Here's](https://github.com/pyinstaller/pyinstaller/runs/3512922457?check_suite_focus=true) another `hook-matplotlib.backends.py` timeout on Windows Python 3.7. [This line](https://github.com/pyinstaller/pyinstaller/runs/3512922457?check_suite_focus=true#step:11:1900) implies that it's getting stuck on the cleanup. The child process has already sent its response, the parent has received it and sent the signal to shut the child down and the parent then waits for the child to terminate but either it never does or its termination is never signalled back to the parent. Could one of the GUI backends have some event loop running in a background thread which prevents the child process from terminating? In which case would adding a `sys.exit()` call to the end of the `_child.py` kill it?

rokm commented:
> Could one of the GUI backends have some event loop running in a background thread which prevents the child process from terminating?

Hmm, I don't think this is the case; the event loops of UI frameworks typically run in the main thread... and I think in `matplotlib`, event loop is run only when you call `figure.show()`.

And besides, here the failure happened when trying to call `matplotlib.get_data_path()` in a subprocess, which should not involve any backend loading.

> In which case would adding a sys.exit() call to the end of the _child.py kill it?

It probably won't make a difference. Either the child ends up stuck before it reaches the end of execution (e.g. reading command line in `run_next_command` blocks forever for some reason), or it could (probably?) get stuck during python cleanup after the `_child.py` effectively exists. 

What we could try is instead of parent waiting for child to exit forever, it does so with a  timeout and if that fails, it tries to kill the child process.

(As a side note, since test timeout manages to kill the gw02 worker, I suspect that this issue occurs in those occasional instances when the test ends up running in a secondary thread instead of the main one).

bwoodsend commented:
Well the fact that we're seeing this in other pull requests at least shows that it's not a regression caused by this pull request so it's OK to merge this one.
